### ISSUE TITLE
Upgrade ZF components. Zend_Json

### ISSUE THREAD

**Issue by okorshenko** [magento/magento2#9236]
**Description**
Upgrade components from ZF1 to ZF2

As a long term goal, we would like to eliminate knowledge about 3rd party libraries from Magento code base. Magento code still can use 3rd party libraries, but they must be wrapped by Magento interfaces and classes (adapters) so that 3rd party libraries can be easily substituted by newest versions or alternative implementations.

**Acceptance Criteria**

1. Magento interfaces are defined and can be used instead of ZF1 classes in Magento code
2. Default implementations for new interfaces are defined and configured as the preferences for new interfaces
3. Client code of the ZF1 components is refactored to use Magento interfaces
4. Zend classes are used as private services in Magento adapters
5. Backward compatibility is preserved
6. ZF1 components are updated to ZF2

**List of the components**

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Json (total: 51)|Framework\DataObject.php|1
&nbsp;|&nbsp;|Framework\Controller|1
&nbsp;|&nbsp;|Framework\Json|2
&nbsp;|&nbsp;|Framework\Module|3
&nbsp;|&nbsp;|Framework\Webapi|3
&nbsp;|&nbsp;|Magento\Weee|6
&nbsp;|&nbsp;|Framework\Data|4
&nbsp;|&nbsp;|Magento\Theme|31

**Comment by okorshenko** [comment]
@dmanners if you would like to work on this, I can assign this to you.

**Comment by dmanners** [comment]
**Updated List of the components**

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Json (total: 15)|Framework\DataObject.php|1
&nbsp;|&nbsp;|Framework\Controller|1
&nbsp;|&nbsp;|Framework\Json|2
&nbsp;|&nbsp;|Framework\Module|3
&nbsp;|&nbsp;|Framework\Webapi|3
&nbsp;|&nbsp;|Framework\Data|4
&nbsp;|&nbsp;|Magento\Widget|1

**Comment by dmanners** [comment]
**Updated List of the components as of July**

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Json (total: 13)|Magento\Customer|1
&nbsp;|&nbsp;|Framework\DataObject.php|1
&nbsp;|&nbsp;|Framework\Controller|1
&nbsp;|&nbsp;|Framework\Json|2
&nbsp;|&nbsp;|Framework\Module|5
&nbsp;|&nbsp;|Framework\Webapi|3

**Comment by dmanners** [comment]
Just noticed I forgot to include the setup module in the list of places to change that has 5 occurrences.

**Comment by dmanners** [comment]
Current status

```
lib/internal/Magento/Framework/Controller/Result/Json.php:        $this->json = \Zend_Json::encode($data, $cycleCheck, $options);
lib/internal/Magento/Framework/Module/Setup/Migration.php:    protected function _jsonDecode($encodedValue, $objectDecodeType = \Zend_Json::TYPE_ARRAY)
lib/internal/Magento/Framework/Module/Setup/Migration.php:        return \Zend_Json::decode($encodedValue, $objectDecodeType);
lib/internal/Magento/Framework/Module/PackageInfo.php:     * @throws \Zend_Json_Exception
lib/internal/Magento/Framework/Module/PackageInfo.php:                        $packageData = \Zend_Json::decode($jsonData[$key]);
lib/internal/Magento/Framework/Module/PackageInfo.php:                    } catch (\Zend_Json_Exception $e) {
lib/internal/Magento/Framework/Module/PackageInfo.php:                        throw new \Zend_Json_Exception(
lib/internal/Magento/Framework/Json/Decoder.php:        return \Zend_Json::decode($data);
lib/internal/Magento/Framework/Json/Encoder.php:        return \Zend_Json::encode($data);
lib/internal/Magento/Framework/Webapi/Rest/Request/Deserializer/Json.php:        } catch (\Zend_Json_Exception $e) {
lib/internal/Magento/Framework/Webapi/Test/Unit/Rest/Request/Deserializer/JsonTest.php:            ->will($this->throwException(new \Zend_Json_Exception));
lib/internal/Magento/Framework/Webapi/Test/Unit/Rest/Request/Deserializer/JsonTest.php:                new \Zend_Json_Exception('Decoding error:' . PHP_EOL . 'Decoding failed: Syntax error')
setup/src/Magento/Setup/Test/Unit/Controller/MarketplaceTest.php:            ->will($this->returnValue(\Zend_Json::encode(['success' => true])));
setup/src/Magento/Setup/Test/Unit/Controller/MarketplaceTest.php:            ->will($this->returnValue(\Zend_Json::encode(['success' => true])));
setup/src/Magento/Setup/Model/PackagesAuth.php:                return \Zend_Json::encode(['success' => true]);
setup/src/Magento/Setup/Model/PackagesAuth.php:                return \Zend_Json::encode(['success' => false, 'message' => 'Bad credentials']);
setup/src/Magento/Setup/Model/PackagesAuth.php:            return \Zend_Json::encode(['success' => false, 'message' => $e->getMessage()]);
```

**Comment by dmanners** [comment]
**Updated List of the components as of 28th of July**

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Json (total: 2)|Framework\Json|2

That makes me feel good!

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-69452

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-69451

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-69369

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-67511

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-67510

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-70775

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-70966

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71007

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71034

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71040

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71061

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71060

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71059

**Comment by magento-team** [comment]
Internal ticket to track issue progress: MAGETWO-71062

**Comment by dmanners** [comment]
Small step backwards but no problem 💃 we go again!

\# | Component Name <br/> (total usages count) | Module Name | Usages Count
------------- | ------------- | ------------- | -------------
1|Zend_Json (total: 3)|Framework\Json|2
2|&nbsp;|Framework\Controller|1

**Comment by magento-engcom-team** [comment]
Hi @dmanners 
Thank you for continuous contribution to this component. You almost resolved all issues in that area! 👍 👍 👍 
We are closing the issue, but feel free to proceed with Zend_Json refactoring

**Comment by ishakhsuvarov** [comment]
Hi @okorshenko. Thank you for your report. 
The issue has been fixed in magento/magento2#13137 by @dmanners in 2.3-develop branch
Related commit(s):
 - 41a4cf635fe8f64d911be07b876fcbe27e93c6fb
 - 9263b1d34f53efeddac072d00ab689da0652042d

The fix will be available with the upcoming patch release.

**Comment by sidolov** [comment]
Hi @okorshenko. Thank you for your report. 
The issue has been fixed in magento/magento2#14438 by @dmanners in 2.3-develop branch
Related commit(s):
 - [cdca73bf922e63a4116518b7128295e29452ebb8](https://github.com/magento/magento2/commit/cdca73bf922e63a4116518b7128295e29452ebb8)
 - [6561541750c62b0199b1994b617ccff579e97793](https://github.com/magento/magento2/commit/6561541750c62b0199b1994b617ccff579e97793)

The fix will be available with the upcoming 2.3.0 release.
